psycop/common/cohort_definition.py

```diff
@@ -52,32 +50,52 @@ def get_outcome_timestamps() -> pl.DataFrame:
         ...
 
 
+msg = Printer(timestamp=True)
+
+
 def filter_prediction_times(
-    prediction_times: pl.DataFrame,
+    prediction_times: pl.LazyFrame,
     filtering_steps: Iterable[PredictionTimeFilter],
     entity_id_col_name: str,
+    get_counts: bool = True,
 ) -> FilteredPredictionTimeBundle:
+    """Apply a series of filters to prediction times.
+
+    If prediction_times is a DataFrame, then the filters will be applied eagerly and we can calculate the stepdeltas.
+    If not, then we will apply the filters lazily and we won't be able to calculate the stepdeltas.
+    """
+
     stepdeltas: list[StepDelta] = []
     for i, filter_step in enumerate(filtering_steps):
-        n_prediction_times_before = prediction_times.shape[0]
-        n_ids_before = prediction_times[entity_id_col_name].n_unique()
+        msg.info(f"Applying filter: {filter_step.__class__.__name__}")
+
+        if get_counts:
+            prefilter_prediction_times = prediction_times.collect()
+
         prediction_times = filter_step.apply(prediction_times)
-        print(f"Applying filter: {filter_step.__class__.__name__}")
-        stepdeltas.append(
-            StepDelta(
-                step_name=filter_step.__class__.__name__,
-                n_prediction_times_before=n_prediction_times_before,
-                n_prediction_times_after=prediction_times.shape[0],
-                n_ids_before=n_ids_before,
-                n_ids_after=prediction_times[entity_id_col_name].n_unique(),
-                step_index=i,
-            ),
-        )
+
+        if get_counts:
```

Comment:
```suggestion
        if get_counts:
        prediction_times_after = prediction_times.collect()
            stepdeltas.append(
                StepDelta(
                    step_name=filter_step.__class__.__name__,
                    n_prediction_times_before=prefilter_prediction_times.shape[0],  # type: ignore
                    n_prediction_times_after=prediction_times_after.shape[0],
                    n_ids_before=prefilter_prediction_times[  # type: ignore
                        entity_id_col_name
                    ].n_unique(),
                    n_ids_after=prediction_times_after[
                        entity_id_col_name
                    ].n_unique(),
                    step_index=i,
                ),
            )
```

Reply:
don't know how long it potentially takes to collect, but in case it is time-consuming ^